pkg/controller/clusterdeprovisionrequest/clusterdeprovisionrequest_controller.go

```diff
@@ -0,0 +1,173 @@
+	rLog := log.WithFields(log.Fields{
+		"name":       request.NamespacedName.String(),
+		"controller": controllerName,
+	})
+	// Fetch the ClusterDeprovisionRequest instance
+	instance := &hivev1.ClusterDeprovisionRequest{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			rLog.Debug("clusterdeprovisionrequest not found, skipping")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		rLog.WithError(err).Error("cannot get clusterdeprovisionrequest")
+		return reconcile.Result{}, err
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		rLog.Debug("clusterdeprovisionrequest being deleted, skipping")
+		return reconcile.Result{}, nil
+	}
+
+	if instance.Status.Completed {
+		rLog.Debug("clusterdeprovisionrequest is complete, skipping")
+		return reconcile.Result{}, nil
+	}
+
+	// Generate an uninstall job
+	hiveImage := images.GetHiveImage(rLog)
+	rLog.Debug("generating uninstall job")
+	uninstallJob, err := install.GenerateUninstallerJobForDeprovisionRequest(instance, hiveImage)
+	if err != nil {
+		rLog.Errorf("error generating uninstaller job: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	rLog.Debug("setting uninstall job controller reference")
+	err = controllerutil.SetControllerReference(instance, uninstallJob, r.scheme)
+	if err != nil {
+		rLog.Errorf("error setting controller reference on job: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// Check if uninstall job already exists:
+	existingJob := &batchv1.Job{}
+	err = r.Get(context.TODO(), types.NamespacedName{Name: uninstallJob.Name, Namespace: uninstallJob.Namespace}, existingJob)
+	if err != nil && errors.IsNotFound(err) {
+		rLog.Debug("uninstall job does not exist, creating it")
+		err = r.Create(context.TODO(), uninstallJob)
+		if err != nil {
+			rLog.Errorf("error creating uninstall job: %v", err)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		rLog.Errorf("error getting uninstall job: %v", err)
+		return reconcile.Result{}, err
+	}
+	rLog.Debug("uninstall job exists, checking its status")
+
+	// Uninstall job exists, check its status and if successful, set the deprovision request status to complete
+	if controllerutils.IsSuccessful(existingJob) {
+		rLog.Infof("uninstall job successful, setting completed status")
+		// jobDuration calculates the time elapsed since the uninstall job started for deprovision job
+		jobDuration := existingJob.Status.CompletionTime.Time.Sub(existingJob.Status.StartTime.Time)
+		rLog.WithField("duration", jobDuration.Seconds()).Debug("uninstall job completed")
+		hivemetrics.MetricUninstallJobDuration.Observe(float64(jobDuration.Seconds()))
```

Comment:
I would report the metric right after you successfully update the request, this would prevent duplicating it in the event we can't save for some reason.

Reply:
ack, makes sense